proxy/proxy_retries_test.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) DataStax, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/datastax/cql-proxy/proxycore"
+	"github.com/datastax/go-cassandra-native-protocol/frame"
+	"github.com/datastax/go-cassandra-native-protocol/message"
+	"github.com/datastax/go-cassandra-native-protocol/primitive"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProxy_Retries(t *testing.T) {
+	const idempotentQuery = "SELECT * FROM test.test"
+	const nonIdempotentQuery = "INSERT INTO test.test (k, v) VALUES ('a', uuid())"
+
+	var tests = []struct {
+		query         string
+		response      message.Error
+		errString     string
+		numNodesTried int
+		retryCount    int
+	}{
+		{ // Bootstrapping error
+			idempotentQuery,
+			&message.IsBootstrapping{ErrorMessage: "Bootstrapping"},
+			"cql error: ERROR UNAVAILABLE (code=ErrorCode Unavailable [0x00001000], msg=No more hosts available (exhausted query plan), cl=ConsistencyLevel ANY [0x0000], required=0, alive=0)",
+			3,
+			2,
+		},
+		{ // Bootstrapping error w/ non-idempotent query
+			nonIdempotentQuery,
+			&message.IsBootstrapping{ErrorMessage: "Bootstrapping"},
+			"cql error: ERROR UNAVAILABLE (code=ErrorCode Unavailable [0x00001000], msg=No more hosts available (exhausted query plan), cl=ConsistencyLevel ANY [0x0000], required=0, alive=0)",
+			3,
+			2,
+		},
+		{ // Error response (truncate), retry until succeeds or exhausts query plan
+			idempotentQuery,
+			&message.TruncateError{ErrorMessage: "Truncate"},
+			"cql error: ERROR UNAVAILABLE (code=ErrorCode Unavailable [0x00001000], msg=No more hosts available (exhausted query plan), cl=ConsistencyLevel ANY [0x0000], required=0, alive=0)",
```

Comment:
This should probably not verify against the exact error string. This is a big fragile.